torch/csrc/jit/pybind_utils.h

```diff
@@ -174,25 +243,67 @@ inline TypedIValue toTypedIValue(py::handle input) {
   }
 }
 
+inline IValue toIValue(
+    py::handle obj,
+    const TypePtr& type,
+    c10::optional<int32_t> N = c10::nullopt);
+
+inline bool isTraceableType(TypePtr type) {
+  if (type->isSubtypeOf(TensorType::get())) {
+    return true;
+  }
+
+  if (auto list_type = type->cast<ListType>()) {
+    return isTraceableType(list_type->getElementType());
+  }
+
+  if (auto tuple_type = type->cast<TupleType>()) {
+    return std::all_of(
+        tuple_type->elements().begin(),
+        tuple_type->elements().end(),
+        [](TypePtr element_type) { return isTraceableType(element_type); });
+  }
+
+  if (auto dict_type = type->cast<DictType>()) {
+    return isTraceableType(dict_type->getValueType());
+  }
+
+  return false;
+}
+
+inline TypedIValue toTraceableIValue(py::handle input) {
+  auto match = tryToInferType(input);
+  if (!match.type) {
+    AT_ERROR(
+        "Tracer cannot infer type of ", py::str(input), "\n:", match.errMsg);
+  }
+  auto type = *match.type;
+
+  if (isTraceableType(type)) {
```

Comment:
Not sure exactly, but I wonder if we can get away with not checking the types here since the values are checked at
https://github.com/pytorch/pytorch/blob/8c57ce87b019ea1c29c4cd8cf69a6170f462bcc2/torch/csrc/jit/tracer.cpp#L215

Still, I think it is fine in this PR and we can look into removing as a follow up.